lib/rust/callback/src/lib.rs

```diff
@@ -189,114 +81,160 @@ impl<Args, F: ?Sized + FnMut<Args>> RegistryFnCall<Args> for RegistryFnMut<F> {
 // === Registry ===
 // ================
 
-/// The main callback registry structure. The [`F`] parameter is either instantiated to
-/// [`RegistryFn<dyn Fn<Args>>`] or to [`RegistryFnMut<dyn FnMut<Args>>`]. See the generated aliases
-/// for common types below, in the [`registry`] module.
+/// The main callback registry structure. The [`F`] parameter is always instantiated to `dyn
+/// FnMut<...>`. See the generated aliases for common types below, in the [`registry`] module.
 #[derive(CloneRef, Derivative)]
 #[derivative(Clone(bound = ""))]
 #[derivative(Debug(bound = ""))]
 #[derivative(Default(bound = ""))]
 #[allow(missing_docs)]
-pub struct Registry<F> {
+pub struct Registry<F: ?Sized> {
+    model: Rc<RegistryModel<F>>,
+}
+
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+#[derivative(Default(bound = ""))]
+#[allow(missing_docs)]
+struct RegistryModel<F: ?Sized> {
+    is_running:               Cell<bool>,
     #[derivative(Debug = "ignore")]
-    callback_list: Rc<RefCell<Vec<(Guard, F)>>>,
+    callback_list:            RefCell<Vec<(Guard, Box<F>)>>,
+    /// Callback list used to register new callbacks when this registry is running. During run,
+    /// the [`callback_list`] is borrowed and thus can't be mutated.
+    #[derivative(Debug = "ignore")]
+    callback_list_during_run: RefCell<Vec<(Guard, Box<F>)>>,
 }
```

Comment:
Some additional information about how these two lists are managed and operated during a run might be good here. 
After reading the code, I'd suggest the following change: 

```suggestion
    /// Temporary buffer to store new callbacks that are registered while the registry is running. During a run,
    /// the [`callback_list`] is borrowed and thus new callbacks cannot be added because it cannot be mutated. 
    /// The buffer is processed and emptied after the registry has finished processing the existing callbacks.
    #[derivative(Debug = "ignore")]
    new_callback_buffer: RefCell<Vec<(Guard, Box<F>)>>,
}
```

Reply:
I added the comment, but I did not change the var name. I feel that `callback_list_during_run` is more descriptive.